### PR TITLE
Fix bug for sample_weights being specialized to None

### PR DESCRIPTION
**Pull request by a6802739**
@anj-s, reference to #24226.

First, Thanks for your help, I'm not sure I was in the right direction, please take a look, and I'll write some test about it then.

### CODE REVIEW

**Comment by anj-s** [comment]
Thanks for the PR! The change looks to be correct. Can you add a correctness test for this? Thanks!

**Comment by a6802739** [comment]
@anj-s , Thanks for your feedback, I have changed some logic in this PR, and I think I met some problems to add a test here, could you give me some reference?

Here is My original thought, I couldn't find a efficient way to verify after calling function `_prepare_feed_values`, the return `ins` contain the `sample_weights`, is there some different behaviour result  the not none sample_weights will lead to? and then I could only check whether the result is match our expectation.

**Comment by anj-s** [comment]
Sorry for the late response. There are a couple of things you need in order to check correctness:
1. A model where you can initialize weights such that output = input (y = x). 
2. Set a slice of your output labels to not be equal to your input data.
3. Let your sample weights array zero out all the indices where the output is not equal to the input.
4. If sample_weights works as expected you should see a 0 loss otherwise your loss will be > 0. The idea being sample weights will zero out all the samples where the output is not equal to the input. 
Let me know if this helps. Thanks!

File: tensorflow/python/keras/distribute/distributed_training_utils.py [hunk 1]
```diff
@@ -575,10 +575,13 @@ def _prepare_feed_values(model, inputs, targets, sample_weights, mode):
   if mode == ModeKeys.PREDICT:
     sample_weights = []
     targets = []
-  else:
+  elif sample_weights is None:
     sample_weights = [
```
Comment (anj-s) [inline tensorflow/python/keras/distribute/distributed_training_utils.py]: I think we need this to be inside the else condition?

**Comment by a6802739** [comment]
@anj-s, Thanks for you kind help, appreciate it very much. I have add the test within `test_sample_weights`, I'm not sure whether it match your saying above, please take a look at it, and then I will try to pass  the unit test.

File: tensorflow/python/keras/distribute/distributed_training_utils.py (hunk 1 above)
Comment (a6802739) [reply tensorflow/python/keras/distribute/distributed_training_utils.py]: Hmm, Sorry, I'm not sure I understand what you mean above. But I have changed the logic there, please take a look at it.

File: tensorflow/python/keras/distribute/distribute_strategy_test.py [hunk 2]
```diff
@@ -689,6 +696,24 @@ def test_numpy_with_sample_weights(self, distribution):
...
+        loss = 'mse'
+        model.compile(optimizer, loss)
+      
+      inputs = np.array([[0, 1, 2, 3],[1, 3, 4, 5]], np.float32)
+      targets = np.array([[0, 1, 2, 1],[1, 2, 2, 3]], np.float32)
+      sample_weights = np.array([[1, 1, 1, 0],[1, 0, 0, 0]], np.float32)
+
+      result = model.evaluate(inputs, targets, sample_weight=sample_weights, 
+                              verbose=1)
+      self.assertAllEqual(result[0], 0)
```
Comment (anj-s) [inline tensorflow/python/keras/distribute/distribute_strategy_test.py]: This looks good. Can you also test the case when you don't pass sample_weights?

File: tensorflow/python/keras/distribute/distributed_training_utils.py (hunk 1 above)
Comment (anj-s) [reply tensorflow/python/keras/distribute/distributed_training_utils.py]: Sorry I think I did not see the latest change for some reason.

File: tensorflow/python/keras/distribute/distributed_training_utils.py [hunk 3]
```diff
@@ -575,10 +575,13 @@ def _prepare_feed_values(model, inputs, targets, sample_weights, mode):
   if mode == ModeKeys.PREDICT:
     sample_weights = []
     targets = []
+  elif sample_weights is not None:
+    sample_weights = training_utils.standardize_sample_weights(sample_weights, model.outputs)
   else:
     sample_weights = [
         None for _ in range(len(model.outputs) * strategy.num_replicas_in_sync)
     ]
+
```
Comment (anj-s) [inline tensorflow/python/keras/distribute/distributed_training_utils.py]: remove extra line?

File: tensorflow/python/keras/distribute/distribute_strategy_test.py [hunk 4]
```diff
@@ -83,7 +83,6 @@ def call(self, inputs):
 
   return _SimpleMLP(num_labels)
 
-
```
Comment (anj-s) [inline tensorflow/python/keras/distribute/distribute_strategy_test.py]: any reason to remove a line here?

File: tensorflow/python/keras/distribute/distributed_training_utils.py (hunk 1 above)
Comment (a6802739) [reply tensorflow/python/keras/distribute/distributed_training_utils.py]: Can you see now?

File: tensorflow/python/keras/distribute/distribute_strategy_test.py (hunk 2 above)
Comment (a6802739) [reply tensorflow/python/keras/distribute/distribute_strategy_test.py]: How about now?

File: tensorflow/python/keras/distribute/distributed_training_utils.py (hunk 3 above)
Comment (a6802739) [reply tensorflow/python/keras/distribute/distributed_training_utils.py]: Done, Thanks.

File: tensorflow/python/keras/distribute/distribute_strategy_test.py (hunk 4 above)
Comment (a6802739) [reply tensorflow/python/keras/distribute/distribute_strategy_test.py]: My mistake. Done now, Thanks.

File: tensorflow/python/keras/distribute/distribute_strategy_test.py [hunk 5]
```diff
@@ -689,6 +697,28 @@ def test_numpy_with_sample_weights(self, distribution):
       model.fit(inputs, targets, sample_weight=sample_weights, epochs=1,
                 steps_per_epoch=2, verbose=1)
 
+  @combinations.generate(
+      combinations.combine(distribution=strategies_minus_tpu, mode=['graph']))
+  def test_sample_weights(self, distribution):
+    with self.cached_session():
+      with distribution.scope():
```
Comment (anj-s) [inline tensorflow/python/keras/distribute/distribute_strategy_test.py]: nit: this can be inlined with the previous with statement.

**Review by anj-s — CHANGES_REQUESTED** [review]
Looks good. I think you need to resolve a few conflicts.

File: tensorflow/python/keras/distribute/distribute_strategy_test.py (hunk 5 above)
Comment (a6802739) [reply tensorflow/python/keras/distribute/distribute_strategy_test.py]: Sorry, I couldn't quite understand what do you mean above, could you please give me some more specific explanation. Thanks.

**Comment by a6802739** [comment]
@anj-s, Thanks for you kind help. Already tackle the conflicts.

File: tensorflow/python/keras/distribute/distribute_strategy_test.py (hunk 5 above)
Comment (anj-s) [reply tensorflow/python/keras/distribute/distribute_strategy_test.py]: nit:I think you can say:
"with self.cached_session(), distribution.scope():"

File: tensorflow/python/keras/distribute/distributed_training_utils.py [hunk 6]
```diff
@@ -645,6 +645,8 @@ def _prepare_feed_values(model, inputs, targets, sample_weights, mode):
     targets = []
   elif not is_distributing_by_cloning(model):
     sample_weights = None  # b/129503665
+  elif sample_weights is not None:
+    sample_weights = training_utils.standardize_sample_weights(sample_weights, model.outputs)
   else:
     sample_weights = [
```
Comment (anj-s) [inline tensorflow/python/keras/distribute/distributed_training_utils.py]: Should sample_weights just be equal to None here? I think with the latest change we don't need to set sample_weights to an array of Nones.

Comment (anj-s) [inline tensorflow/python/keras/distribute/distributed_training_utils.py]: Should sample_weights just be equal to None here? I think with the latest change we don't need to set sample_weights to an array of Nones.

**Review by anj-s — CHANGES_REQUESTED** [review]

File: tensorflow/python/keras/distribute/distribute_strategy_test.py (hunk 5 above)
Comment (a6802739) [reply tensorflow/python/keras/distribute/distribute_strategy_test.py]: I see, Thanks, just want to be consistent with the test above/below. Now I have changed it.

File: tensorflow/python/keras/distribute/distributed_training_utils.py (hunk 6 above)
Comment (a6802739) [reply tensorflow/python/keras/distribute/distributed_training_utils.py]: hmm, I think for distributed scenario, it should be an  array of Nones.

Comment (anj-s) [reply tensorflow/python/keras/distribute/distributed_training_utils.py]: I think a recent change https://github.com/tensorflow/tensorflow/commit/41abdd1212777a60e0d97ada4a39d13f2b9014a1 does not require us to set sample_weights to an array of Nones if sample_weights is not set by the user.  Can you set it to None and see if that works? I think None is the default so maybe that is not necessary as well.

File: tensorflow/python/keras/distribute/distribute_strategy_test.py [hunk 7]
```diff
@@ -740,6 +748,28 @@ def test_numpy_with_sample_weights(self, distribution, cloning):
       model.fit(inputs, targets, sample_weight=sample_weights, epochs=1,
                 steps_per_epoch=2, verbose=1)
 
+  @combinations.generate(
+      combinations.combine(distribution=strategies_minus_tpu, mode=['graph']))
```
Comment (anj-s) [inline tensorflow/python/keras/distribute/distribute_strategy_test.py]: Any reason we did not use "distribution=strategy_minus_tpu_combinations" to test "graph" and "eager"?

**Review by anj-s — CHANGES_REQUESTED** [review]

File: tensorflow/python/keras/distribute/distribute_strategy_test.py (hunk 7 above)
Comment (anj-s) [reply tensorflow/python/keras/distribute/distribute_strategy_test.py]: If we run into issues while testing mode="eager" we can look into it in a followup PR. For now it will be good to submit what we have.

File: tensorflow/python/keras/distribute/distributed_training_utils.py [hunk 8]
```diff
@@ -645,6 +645,8 @@ def _prepare_feed_values(model, inputs, targets, sample_weights, mode):
     targets = []
   elif not is_distributing_by_cloning(model):
     sample_weights = None  # b/129503665
```
Comment (isaprykin) [inline tensorflow/python/keras/distribute/distributed_training_utils.py]: Hey @a6802739!  Thanks for making this change, it's important for us.

I was passing by and noticed that you were trying to make the eager case work.  I think in Eager `sample_weights` at this point should `PerReplica` objects.  Right now they are overwritten with `None`.  I think instead those PerReplica objects needs to be directly passed into `ins` without being overwritten.   That's what it seems to me.  I hope this helps and good luck!

**Comment by anj-s** [comment]
The tests on this PR fail due to more recent changes to the codebase. Closing this PR since the fix for this issue has been submitted.
https://github.com/tensorflow/tensorflow/commit/98a0c57c44d9c4f1a1ad2f7003c83dbe2802f7ce
